packages/embark/src/lib/modules/console/index.ts

```diff
@@ -187,19 +206,23 @@ class Console {
           // TODO add docs link to how to install one
           this.logger.warn(__("If you did not install a blockchain connector, stop this process and install one"));
         }, 5000);
-        this.events.once("blockchain:connector:ready", () => {
+        const clearTimeoutAndContinue = () => {
           clearTimeout(waitingForReady);
-          next();
-        });
+          if (typeof next === "function") {
+            next();
+          }
+        };
+        this.events.request("blockchain:connector:ready", clearTimeoutAndContinue);
+        this.events.once("blockchain:connector:ready", clearTimeoutAndContinue);
```

Comment:
The reason there are both an `events.request` and an `events.once` for the same event is due to the loading order of the plugins.

In the normal run/console situation, the plugins (and therefore the `web3Connecor`) run first, and thus establish an `events.registerCommandHandler("blockchain:connector:ready")`, which suits the `events.request("blockchain:connector:ready")`. 

However, in the test situation, the `Console` loads **AFTER** the plugins, and thus the `events.registerCommandHandler("blockchain:connector:ready")` has not run yet, hence the need to listen for the `blockchain:connector:ready` event.

Reply:
What I've been thinking for awhile is to have retroactive `requests`. By that I mean, if you do like you said `request("blockchain:connector:ready")` but the event doesn't exist yet, you're doomed. The event won't ever fire. 
So it'd be cool to have the events manager store the `requests` and when a new `setCommandHandler` is created, we check to see if one of the request is fullfilled by it. When it does, we remove the request from the list.

That's obviously a more complex solution to your problem, but I feel it would solve a lot of headaches in the future.

For a more simple solution, we could always add an event  on the plugin manager (`plugins.js`) that's called something like `plugin:ready` and tells us when the plugins are all registered and we can actually call our request for the connector after.

Reply:
That's a great idea, I think it would simplify a lot of the headache with trying to get different modules loaded at the right time. Instead we just load modules async as fast as possible and the events manager would take care of firing the event as you mentioned.

Building on the events manager idea, we could even add `async.auto` for event requests so that a particular event only fires once all of its conditions are met. For example, `console:ready` could depend on `runcode:ready` and `blockchain:ready`, and so any event requests for `console:ready` wouldn't fire until those dependent events have also fired.

For the more simple solution mentioned, I actually tried this, but quickly ran in to the issue that the plugin loading does not have a callback and so we never actually know when the plugins are finished loading unfortunately. We could add in an `init(cb)` function that each plugin would need to have (could also put it in an `IPlugin`interface to enforce the method to be present.

Reply:
@jrainville - added the event manager "oracle": https://github.com/embark-framework/embark/pull/1340. This PR is now rebased (and based) on that branch.